### ISSUE TITLE
Practitioner: Generate Entry of Appearance Document for E-Filing

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8455]
As a practitioner, so that I can easily file entries of appearance, I need a workflow that assembles and files an entry of appearance form for me.

The form for an Entry of Appearance (EA) form largely contains information already known by DAWSON; in order to make filing these as easy as possible, we would like to give practitioners an option to simply indicate the party(s) on whose behalf they are making an appearance, then have DAWSON assemble the completed form for approval and filing. This only applies to EA documents; Limited Entry of Appearance and any other documents that appear as selections in the Request Access to This Case workflow are **not** in scope.

To begin, we will implement this without the ability to multidocket these forms across a consolidated group.

## Acceptance Criteria
* When a practitioner invokes the "Request Access to Case" workflow, they are given the option to generate and file an Entry of Appearance
* Private practitioners must indicate the party(s) on whose behalf they are making an appearance, after which the document is generated
* DAWSON generates the EA document using values from the Case and practitioner user record and presents it to the user for approval
  - If the practitioner approves the document, proceed to the Review Your Filing screen as normal to complete the e-filing flow using the generated EA
  - If the practitioner does not approve the document, return them to the original Request Access to This Case screen
* Sample document (current) and desired document (DAWSON-generated) along with data points attached in Notes
* This feature is available to all practitioners, however the auto-generated form may *not* be multidocketed by *any* practitioner.

## Notes
[https://app.zenhub.com/files/152320868/a8581d06-4690-4db4-ae57-c4dd9c852444/download](https://app.zenhub.com/files/152320868/a8581d06-4690-4db4-ae57-c4dd9c852444/download)   

[https://app.zenhub.com/files/152320868/97354ce2-82d1-4fc4-911a-693e5b15872f/download](https://app.zenhub.com/files/152320868/97354ce2-82d1-4fc4-911a-693e5b15872f/download)

* Can/should we take this opportunity to implement the redaction gate on filing in the Request Access to Case flow?

## Story Definition of Ready (updated on 12/23/22)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Is framed in business/user need, the value has been addressed.
- [ ] Includes acceptance criteria
- [ ] Has been refined
- [ ] Pre conditions have been satisfied. 
**Process:** 
Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written, including visual tests for newly added PDFs
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Acceptance criteria for the story has been met
 - [ ] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by katiecissell** [comment]
Pre-refinement questions:
Do we need to handle signature in a special way? is it required? 
Are all these fields required?
Should we let them edit the pre-filled in fields?
Is this a replacement for the form or an additional option to fill this in?
Clarification on sentence "return them to the original Request Access to This Case screen" Original form today or just backing up in the workflow?

**Comment by katiecissell** [comment]
Request access to case page
- Remove the blue background. (If easy, if not, we can keep it.)
- The order of the Entry of Appearance page sections will change, Type of document is first, followed by the tell us about the parties, tell us about the document and selecting extra items to include with the doc sections.
- We will update the redact info reminder alert to the new styling.
- If the entry of appearance document is chosen, then the new radio buttons for generating or uploading the EA pdf will appear below the redaction reminder.
- Auto-generate Entry of Appearance PDF   (default)
- - If they select auto-generate, Dawson will generate the PDF
- Upload PDF form
- - If they select upload, show them the file upload button and the checkboxes for attachments and certificate of service.
- After they have made their selections and click review filing, generate the PDF if applicable
<s>- - Use the 0-100% blue popup, follow the pattern that Cody did with his when we updated it to show the actual progress. The words should say “Generating PDF... </s>

EA for Petitioner - Generate
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBamNGIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--38036a63cb046f3a4377f8c79bcc70f53e81e91b/image.png)


EA for Petitioner - upload
![image.png](https://images.zenhubusercontent.com/63ecf8daa45ad187539a9d61/6c011242-4478-43fa-988a-45f4a68098fc)

EA for Respondent - Generate
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBajBGIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2fa8580a76f978e6a32dea4d8e23867c467616a4/image.png)





EA for Respondent - Upload
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBajhGIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2d65d699b9a116800e58c6fe1ff51421d367713e/image.png)


”Review your filing”
- If we are generating the PDF, on the right side of the review your filing page, display a Preview of the PDF Dawson generated. (If they uploaded, we don’t show the PDF preview)
- Shrink the existing cards (and their contents) so they will fit in the space. It looks like the preview is more than half but less than 2/3 of the page? Maintain that spacing that we see in other areas of the app like Petitions QC or document view
- We will update the redact info reminder alert to the new styling. 
- If a PDF is uploaded, add the redaction checkbox above the submit filing button. (We don’t need the checkbox if we’re generating the PDF)
- when the Submit filing is clicked, show the loading spinner we already have in place today.

Review - Practitioner Generated
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBa0FGIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--8cf47f170e9505302fa9475e85fc75c46578d612/image.png)


Review - Practitioner uploaded
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBdHdaIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--b99c46f28655d3439359a6091820b91a8677abc3/image.png)



PDFS
-The PDF we generate should look like the screenshots below (spacing doesn’t have to be exact as in the mocks but the info displayed is)
If there are multiple lines of address, display them as we do in case information.
If phone number is n/a, do not display anything.
If there are multiple petitioners being represented, put comma's between them until the last one, which will have a & in front of it.
Petitioner
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbFVIIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--ded2bf11fe3d6033d32a27c46529083792e49087/image.png)


Petitioner estate
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBcW1tIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--6ea36cd63e3235800618a79ec2ac9c3f18f7ab9b/image.png)



Respondent
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBaElIIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--fd66a7b91e0de47e872874c4173444d370fb6340/image.png)



- Both Petitioner and Respondent updated UIs should work with Mobile devices
- The PDF preview will not appear on Mobile, the user will have to open the PDF by clicking the PDF link if they want to view it.
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBa01GIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--afb8c8a50d54cf30f3631e95fcae12e092180750/image.png)

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBa1FGIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--9d80190d0b0f86d50cf59e8caa3a9010a224c5f7/image.png)

**Comment by ttlenard** [comment]
## Test Cases


#### 1) Private Practitioner requests access to a case; Redaction text on the page is now displayed with a yellow background.

- Log in as a Private Practitioner
- On the My Cases Page, type in a docket number in the Docket Search
- Click on the Request Access to Case button
- Review the Request Access to Case page

Expected Results:
- [x] **NEW FUNCTIONALITY** - The Redaction reminder text below the "Tell Us About This Document" Section now has a yellow background

*Repeat this test with an IRS Practitioner.



#### 2) Private Practitioner selects all options besides Entry of Appearance from the document type dropdown; no option for Auto-generated document is available.

- Continuing from the previous test
- Click on the Dropdown menu for document type and select a value besides Entry of Appearance

Expected Results:
- [x] There is no option to select to auto-generate the Entry of Appearance
  
*Repeat this test with an IRS Practitioner



#### 3) Private Practitioner selects Entry of Appearance from the document type dropdown; New radio button option is available to Auto-Generate the Entry of Appearance PDF.

- Continuing from the previous test
- Select Entry of Appearance from the Document Type dropdown 
- Review the page

Expected Results:
- [x] **NEW FUNCTIONALITY** - There are two radio buttons under the Tell Us About This Document section
- [x] **NEW FUNCTIONALITY** - One radio button displays "Auto-generate Entry of Appearance PDF (Use only if you do not need to add attachments or a Certificate of Service)
- [x] **NEW FUNCTIONALITY** - One radio button displays Upload PDF Form
- [x] **NEW FUNCTIONALITY** - User can select one or the other radio buttons, but not both
- [x] **NEW FUNCTIONALITY** - When the Upload PDF Form radio button is selected, a section appears for the user to choose a file and they have the option to select attachments or C/S to include with the document.
- [x] **NEW FUNCTIONALITY** - For Private Practitioner filing an EA on a case in a consolidated group, the "Which cases do you want to File in?" section is NEVER available.
- [x] Review Filing button and cancel link is available for either selection.
- [x] For Private Practitioner only, there is a section for "Tell us about the Parties You're Representing".

*Repeat this test with an IRS Practitioner
- [x] **NEW FUNCTIONALITY** - For IRS only filing an EA on a case in a consolidated group, when the Upload PDF form radio button is selected, the section "Which Cases Do you Want To File In?" is available and displays the cases appropriately.
- [x] **NEW FUNCTIONALITY** - For IRS filing an EA on a case in a consolidated group, when the Auto-generate PDF radio button is selected, the "Which Cases Do You Want to File In?" is NOT displayed.


#### 4) Private Practitioner selects Auto-generate Entry of Appearance PDF, Clicks Review Filing.

- Continuing from the previous test
- Be sure to select the radio button to Auto-generate the EA PDF
- Click Review Filing
- Review the "Review Your Filing" screen:

Expected Results:
- [x] **NEW FUNCTIONALITY** - The Redaction reminder text below the Review Your Filing now has a yellow background
- [x] **NEW FUNCTIONALITY** - There is a document preview on the right side of the screen where the user can review the auto-generated document
- [x] **NEW FUNCTIONALITY** - Auto-generated document matches the format in the sample forms/forms in mockups
- [ ] **NEW FUNCTIONALITY** - If the form was generated for a private practitioner, the party/parties that they selected in the previous step are populated in the form correctly
- [x] **NEW FUNCTIONALITY** - The Name, Address (1-3 address lines depending on what is stored for the practitioner), City, State, Phone, Bar #, and email address are all populated correctly based off of the information that is stored for the Practitioner in the Database (You will need to review the information about the practitioner in the DB to ensure that the data matches.)
- [x] **NEW FUNCTIONALITY** - The format of the auto-generated EA matches the mockups
- [x] Clicking on the Document link displays the same auto-generated EA in a pop-up window.
- [x] There is no section for "Document Includes" displayed for auto-generated EA's.
- [x] There is no checkbox present for the user to certify that they have redacted information from their filing

*Repeat this test with an IRS Practitioner
- [x] **NEW FUNCTIONALITY** - If the form was generated for an IRS Practitioner, the party information states "Respondent"


*Be sure to test this with a practitioner that has an international address to ensure that their address information appears on the Auto-Generated EA correctly



#### 5) Private Practitioner submits the auto-generated EA document.

- Continuing from the previous test
- Click Submit Your Filing

Expected Results:
- [x] **NEW FUNCTIONALITY** - Generating PDF progress window is displayed
- [x] **NEW FUNCTIONALITY** - After PDF is generated, user is navigated back to the case docket record.
- [x] **NEW FUNCTIONALITY** - Green confirmation banner with a Print Receipt link
- [x] **NEW FUNCTIONALITY** - After clicking on the print receipt, Receipt of Filing opens.  Information is displayed appropriately ex: Filed by lists the practitioner name, Date filed info is appropriate, and the EA is titled appropriately 
- [x] **NEW FUNCTIONALITY** - The Entry of Appearance document is displayed on the docket record
     - [x] Filed date is correct
     - [x] Event code is EA
     - [ ] Document title reflects the petitioners that you selected appropriately
     - [x] Page number is correct (Should be 2 pages, 1 page is the EA, one page is the coversheet)
     - [x] Filed by is the practitioner name
     - [x] Served date is today's date
     - [x] Served party code is B
- [x] **NEW FUNCTIONALITY** - After clicking on the link to view the document, the document opens in a new tab
     - [x] Coversheet is appended to the auto-generated document
     - [x] Coversheet has the correct case caption, electronically filed, correct docket number, and document number, correct title, and correct served date.
     - [x] Auto-generated EA matches the format in the mockup
- [x] Service logs in Kibana reflect service was sent appropriately

*Repeat this test with an IRS Practitioner
- [x] **NEW FUNCTIONALITY** - If the form was generated for an IRS Practitioner, the Filed by on the docket record should say "Resp."


#### 6) Docket Clerk QC's the auto-generated EA document.

**Part 1**
- Continuing from the previous test
- log in as a docket clerk
- Click on Document QC
- Click on Section Document QC
- Find the Auto-generated EA document that was just submitted in the list of work items
- Click on the document title to QC the document

Expected Results
- [x] Document Title is correctly displayed
- [x] Filed date/Served date is correctly displayed
- [x] Date received is correct
- [x] Document type is Entry of Appearance
- [x] No additional info 1, Additional info 2, or inclusions are populated. The add to coversheet and the Add to Pending report is not checked
- [x] Filed by displays the correct Practitioner info.
- [x] Document in the document preview has the coversheet and matches the mockups

**Part 2**
- Click on Complete

Expected Results:
- [x] Green Confirmation banner
- [x] QC item appears correctly in the processed tab of your QC

*Repeat this test by also clicking complete and send message
*Repeat this test by also QCing an auto-generated EA from an IRS Practitioner



#### 7) Private Practitioner uploads and then submits the uploaded EA document

**Part 1**
- Log in as a Private Practitioner
- Search for a case docket number
- Click on Request Access to Case button
- Select Entry of Appearance
- Be sure that you click the radio button for Upload PDF Form
- Click on the Choose File button and upload a PDF
- Check the box for Attachments and Certificate of Service
- Select the Party that you are representing
- Click on Review Filing

Expected Results:
- [x] **NEW FUNCTIONALITY** - Redaction text at the top of the screen now has a yellow background
- [x] **NEW FUNCTIONALITY** - The section with the checkbox to certify redaction has now been added.
- [x] User cannot check submit the filing until the redaction acknowledgement checkbox has been checked

**Part 2**
- Click on the Submit Filing button

Expected Results:
- [x] Uploading document progress bar appears
- [x] Once it is complete, user is navigated back to the docket record
- [x] Green confirmation banner with print receipt link is displayed
- [x] Regression test to ensure that the filing displays appropriately on the docket record, and that service was sent.


*Repeat this test as an IRS Practitioner.



#### 8) Docket Clerk QC's the uploaded EA document

- This is a regression test
- Log in as a docket clerk
- Click on Document QC
- Click on Section Document QC
- Find the EA that you just submitted
- Click on the Document Title

Expected Results:
- [x] Docket clerk can QC the uploaded document as normal.

*Be sure that you can QC an uploaded EA from both a Private Practitioner AND an IRS practitioner



#### 9) IRS Respondent submits an uploaded EA document and files it across the group of consolidated cases; document is filed in each case appropriately

- This is mostly a regression test
- Log in as an IRS Respondent
- Search for a case that is part of a consolidated group of cases
- Click on the Request Access to Case button
- Select Entry of Appearance
- Be sure that you click the radio button for Upload PDF Form
- Click on the Choose File button and upload a PDF
- Check the box for Attachments and Certificate of Service
- Select the radio button to file the EA in all cases in the group
- Click on Review Filing

Expected Results:
- [x] **NEW FUNCTIONALITY** - Redaction text at the top of the screen now has a yellow background
- [x] Tiles for cases this document will be filed in and Service parties are displayed
- [x] **NEW FUNCTIONALITY** - The section with the checkbox to certify redaction has now been added.
- [x] User cannot check submit the filing until the redaction acknowledgement checkbox has been checked

**Part 2**
- Click on the Submit Filing button

Expected Results:
- [x] Uploading document progress bar appears
- [x] Once it is complete, user is navigated back to the docket record
- [x] Green confirmation banner indicating that the document was filed in all selected cases in the group with print receipt link is displayed
- [x] Print receipt lists all of the cases
- [x] Regression test to ensure that the filing displays appropriately on the docket record (with the appropriate coversheet), and that service was sent to all parties in the cases in the group.



#### 10) Docket Clerk QC's the Uploaded EA document that was filed across the group

- This is a regression test
- Log in as a docket clerk
- Click on Document QC
- Click on Section Document QC
- Find the EA's that you just submitted across the group
- Click on the Document Titles

Expected Results:
- [x] Docket clerk can QC the uploaded documents that were filed across the group as per normal.



#### 11) Complete the Practitioner tests above on a mobile device

- [ ] User can complete the actions on mobile and have a good user experience
- [ ] **NEW FUNCTIONALITY** - The document preview on the generated EA document will not display on mobile, but the user can click on the hyperlink to preview the generated document prior to submitting.

**Comment by ttlenard** [comment]
Some testing feedback:

If the user takes more than 2 minutes (I think) on the Review your Filing page to click on the Document preview link:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBb2ZUIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--07833ee328f9b5b43723d0095ee521254f9b9cf8/image.png)

They will get this error message:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBb3JUIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--49d6832e3af7167a99599d2f049b3b10d86736c6/image.png)

And then when they attempt to submit the EA to the Court, they will get the Uh-oh error message with this error:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBb3ZUIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--b4d72ae5f483e988293a8480754728b340044f43/image.png)

**Comment by ttlenard** [comment]
@TomElliottFlexion @zRogersFlexion @codyseibert 

Some more testing feedback:
The "All Done" message is showing very briefly before the "Preparing upload" message.
- Log in as a practitioner (I was IRS)
- Search for a docket number that you don't yet have access to
- request access to case
- Entry of Appearance
- Generate it
- Submit the Filing
- Watch very closely - the All Done Message pops up very briefly and then the Preparing upload message appears.  

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbnpiIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--10a6bb4e4bcb31679264c152befe5c155296b0a5/image.png)
